lib/matplotlib/widgets.py

```diff
@@ -1833,3 +1833,90 @@ def onmove(self, event):
             self.canvas.blit(self.ax.bbox)
         else:
             self.canvas.draw_idle()
+
+
+class Sketcher(object):
+    """
+    Tool to sketch on the axes.
+
+    Parameters
+    ----------
+    ax : Axes
+        The axes to sketch on
+
+    """
+    def __init__(self, ax):
+        self.ax = ax
+        self.point_cache = []
+        self.cid_cache = {}
+        self._active = False
+        self.active = True
+        self.lines = []
+
+    def _safe_connect(self, msg, func):
+        """
+        Connect a signal, but make sure we disconnect the existing one first
+        """
+        self._safe_disconnect(msg)
+
+        cid = self.ax.figure.canvas.mpl_connect(msg, func)
+        self.cid_cache[msg] = cid
+
+    def _safe_disconnect(self, msg):
+        """
+        Disconnect a signal, even it is not really connected
+        """
+        old_cid = self.cid_cache.pop(msg, None)
+
+        if old_cid is not None:
+            self.ax.figure.canvas.mpl_disconnect(old_cid)
+
+    def _down_event(self, ev):
+        """
+        button down callback
+        """
+        del self.point_cache[:]
```

Comment:
why clear the cache prior to checking for widgetlock? If the widget doesn't have lock, then it means that the widget wasn't the intended recipient of the signal, right?